eclass/zig-toolchain.eclass

```diff
@@ -0,0 +1,325 @@
+zig-toolchain_get_target() {
+	[[ ${#} -eq 1 ]] || die "${FUNCNAME[0]}: exactly one argument must be passed"
+	local c_target=${1}
+	local c_target_prefix=${c_target%%-*}
+	local c_target_suffix=${c_target##*-}
+
+	local arch os abi
+
+	case ${c_target_prefix} in
+		i?86)	arch=x86;;
+		arm64)	arch=aarch64;;
+		arm*)	arch=arm;;
+		*)	arch=${c_target_prefix};;
+	esac
+
+	case ${c_target} in
+		*linux*)	os=linux;;
+		*apple*)	os=macos;;
+	esac
+
+	case ${c_target_suffix} in
+		solaris*)	os=solaris abi=none;;
+		darwin*)	abi=none;;
+		*)		abi=${c_target_suffix};;
+	esac
+
+	[[ ${arch} == arm && ${abi} == gnu ]] && die "${FUNCNAME[0]}: Zig does not support old ARM ABI"
+
+	echo ${arch}-${os}-${abi}
+}
+
+
+
+# @FUNCTION: zig-toolchain_get_cpu
+# @USAGE: <C-style target triple>
+# @DESCRIPTION:
+# Translates C-style target triple (like CHOST or CBUILD)
+# to Zig-style target CPU and features. Mostly used to get generic target CPU
+# during cross-compiling if user does not set ZIG_CPU variable.
+#
+# See ZIG_CPU description for more information.
+zig-toolchain_get_cpu() {
+	[[ ${#} -eq 1 ]] || die "${FUNCNAME[0]}: exactly one argument must be passed"
+	local c_target=${1}
+	local c_target_prefix=${c_target%%-*}
+
+	local base_cpu features=""
+
+	case ${c_target_prefix} in
+		i?86)					base_cpu=${c_target_prefix};;
+		loongarch64)			base_cpu=generic_la64;;
+		powerpc | powerpcle)	base_cpu=ppc;;
+		powerpc64)				base_cpu=ppc64;;
+		powerpc64le)			base_cpu=ppc64le;;
+		riscv32)				base_cpu=generic_rv32;;
+		riscv64)				base_cpu=generic_rv64;;
+		*)						base_cpu=generic;;
+	esac
+
+	case ${c_target_prefix} in
+		armv5tel)	features+="+v5te";;
+		armv*)		features+="+${c_target_prefix##armv}";;
+	esac
+
+	case ${c_target_prefix} in
+		arm64)	;;
+		arm*)
+			local is_softfloat=$(CTARGET=${c_target} tc-tuple-is-softfloat)
+			case ${is_softfloat} in
+				only | yes) features+="+soft_float";;
+				softfp | no) features+="-soft_float";;
+				*) die "${FUNCNAME[0]}: tc-tuple-is-softfloat returned unexpected value"
+			esac
+		;;
+	esac
+
+	echo ${base_cpu}${features}
+}
+
+# @FUNCTION: zig-toolchain_find_installation
+# @DESCRIPTION:
+# Detects suitable Zig installation and sets ZIG_VER and ZIG_EXE variables.
+#
+# See ZIG_EXE and ZIG_VER descriptions for more information.
+zig-toolchain_find_installation() {
+	# Adapted from https://github.com/gentoo/gentoo/pull/28986
+	# Many thanks to Florian Schmaus (Flowdalic)!
+
+	[[ -n ${ZIG_SLOT} ]] || die "${FUNCNAME[0]}: ZIG_SLOT must be set"
+	if ver_test "${ZIG_SLOT}" -lt "0.13"; then
+		die "${ECLASS}: ZIG_SLOT must be >= 0.13, found ${ZIG_SLOT}"
+	fi
+
+	einfo "Searching Zig ${ZIG_SLOT}..."
+
+	local selected_path selected_version
+
+	# Prefer "dev-lang/zig" over "dev-lang/zig-bin"
+	local -a candidates candidates1 candidates2
+	readarray -d '' -t candidates1 < <(find -P "${BROOT}"/usr/bin/ -mindepth 1 -maxdepth 1 -type l -name "zig-bin-*" -print0 2> /dev/null || echo -n "")
```

Comment:
I'm not a fan of this. Can't we just have a list of known zig slots in the eclass and iterate from high-to-low, checking zig and zig-bin for each?

Reply:
Do you mean something like this?
```baah
local zig_vers=(
    "9999"
    "0.13.1"
    "0.13.0"
)

local sel_ver
for sel_ver in zig_ver; do
   if [[ -z "${BROOT}/usr/bin/zig-${sel_ver} ]]; then
        break
    fi
    if [[ -z "${BROOT}/usr/bin/zig-bin-${sel_ver} ]]; then
        break
    fi
done
```

Reply:
Yep, that looks much better. Thanks.

Reply:
Doesn't it mean that if someone adds zig ebuild in their overlay earlier than `::gentoo`, they would not be able to set `ZIG_SLOT="new_version"` unless they copy eclass for 2-lines change? This version does not have this limitation....

Reply:
I don't see that as a particularly terrible limitation. I really don't like probing `${BROOT}` for whatever binary we happen to find.

Reply:
3767c578f1d7f21637fc19e6d19e407cd2bbf423